File: Notes/05_Django_GraphQL/01_Intro_Install_and_First_Query/books/schema.py
# schema describe our data models which we will provide to our graphQL server

import graphene
from graphene_django import DjangoObjectType
# This "DjangoObjectType" is a way of formatting our django object that can be utilize by our GraphQL
from .models import Books


class BookType(DjangoObjectType):
    # Now here we will just describe the model that we want to use and the data that we want to collect
    # Same as that we do in serializer
    class Meta:
        model = Books
        fields = ("id", "title", "excerpt")
        """
            so, the type that we create above will be translated like this which then GraphQL will create graph qnd that we can query that
            type Books{
                id: id,
                title: String
                Excerpt: String
            }
        """


# Now we need a way to generate a query
class Query(graphene.ObjectType):
    """
    type Query{
        me: User
    }
    """
    # we added the books into query so that we can be able to query all the books on books data
    all_books = graphene.List(BookType)
    # so here we are using 'all_books' which have an access of List of "BookType"
    # now we can use this 'all_books' to query the list of books

    def resolve_all_books(root, info):
        # so here we are resolving the query 'all_books' and we want to perform some specific operation and return the data using this function
        # return Books.objects.all()
        return Books.objects.filter(title="django")


# now we have create the Types and Query out of those type on we will build schema out of it
schema = graphene.Schema(query=Query)

File: Notes/00_Main/server/Topics_wise/07_GraphQL_Schema/schema/schema.js
const graphql = require("graphql");

/*
    -> schema will describe the data on this kind of graph
    -> it describe object types

    -> for this project we will going to create to object types:
        -> books
        -> Authors
*/

const { GraphQLObjectType, GraphQLString } = graphql;

// now we will define new graphql type
const BookType = new GraphQLObjectType({
  // now here we have to define what this book type is all about
  name: "Book",
  //   need to give name
  fields: () => ({
    //   now here we will define all the field related to 'books' object type
    id: {
      type: GraphQLString,
      // we have to use graphql string
    },
    name: {
      type: GraphQLString,
    },
    genre: {
      type: GraphQLString,
    },
  }),
});
